# Assignment 1

In [137]:
import numpy as np
import pandas as pd
import sklearn as sl
import matplotlib as mp
import scipy as cp
import seaborn as sb
from matplotlib import pyplot as plt
import statsmodels.api as sm
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression, Ridge, Lasso
from sklearn.metrics import f1_score, mean_absolute_error, mean_squared_error, mean_absolute_percentage_error
from math import sqrt
from sklearn import linear_model as lm
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import make_scorer

## Part A

1.	Download the “dataset_lm.csv” file from Canvas and upload it to Jupyter Notebook. 

In [138]:
df = pd.read_csv(r"C:\Users\johns\OneDrive\Desktop\MBAN Semester 2\Predictive Analytics\Assignment 1\dataset_lm.csv")

In [139]:
df.head()

Dependent Var  Explanatory Var #1  Explanatory Var #2  Explanatory Var #3  \
0      56.293458           13.698667           50.639873                   0   
1      58.473431            2.714725           65.845845                   1   
2      94.195330           11.618072           65.072497                   0   
3      29.074583            0.818623           45.408996                   1   
4      86.035569            9.077544           73.548021                   0   

   Explanatory Var #4  Explanatory Var #5  Explanatory Var #6  \
0          -18.568035           45.121911           11.412501   
1          -25.105932           47.190213           10.080280   
2           -7.897464           52.163036           11.057301   
3          -18.316132           54.356714            5.029029   
4          -19.204165           47.186807           12.128134   

   Explanatory Var #7  Explanatory Var #8  Explanatory Var #9  \
0           56.410757                   2          -12.281132   
1           65.383107                   3          -36.763585   
2           82.812717                   0          -15.733547   
3           48.812471                   1          -12.825591   
4           62.520911                   2          -13.804860   

   Explanatory Var #10  Explanatory Var #11  Explanatory Var #12  \
0            38.996909            -3.010548            49.195073   
1            51.654939             4.991111            45.591729   
2            48.913837            -2.457696            56.608806   
3            45.851732            14.974177            47.362594   
4            47.765904             9.593982            53.700562   

   Explanatory Var #13  Explanatory Var #14  Explanatory Var #15  
0                    0           -21.153143            46.919314  
1                    0            -6.474403            53.383508  
2                    0           -27.903299            48.515026  
3                    1           -10.064411            55.266254  
4                    0           -17.546302            48.150543

In [140]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 422 entries, 0 to 421
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Dependent Var        422 non-null    float64
 1   Explanatory Var #1   422 non-null    float64
 2   Explanatory Var #2   422 non-null    float64
 3   Explanatory Var #3   422 non-null    int64  
 4   Explanatory Var #4   422 non-null    float64
 5   Explanatory Var #5   422 non-null    float64
 6   Explanatory Var #6   422 non-null    float64
 7   Explanatory Var #7   422 non-null    float64
 8   Explanatory Var #8   422 non-null    int64  
 9   Explanatory Var #9   422 non-null    float64
 10  Explanatory Var #10  422 non-null    float64
 11  Explanatory Var #11  422 non-null    float64
 12  Explanatory Var #12  422 non-null    float64
 13  Explanatory Var #13  422 non-null    int64  
 14  Explanatory Var #14  422 non-null    float64
 15  Explanatory Var #15  422 non-null    flo

2.	Run the OLS model by using the dependent and explanatory variables in the dataset. 

In [141]:
y = df['Dependent Var']
X = df.drop('Dependent Var', axis = 1)

??? do I add a constant - sm.add_constant(x)

add one because of this

Notes:

[1] R² is computed without centering (uncentered) since the model does not contain a constant.

[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.

In [142]:
model_reg = sm.OLS(y,X).fit()

3.	Show your summary table in Python and interpret your results in the summary report.

In [143]:
model_reg.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:          Dependent Var   R-squared (uncentered):                   0.999
Model:                            OLS   Adj. R-squared (uncentered):              0.999
Method:                 Least Squares   F-statistic:                          4.509e+04
Date:                Mon, 23 Oct 2023   Prob (F-statistic):                        0.00
Time:                        21:20:27   Log-Likelihood:                         -841.76
No. Observations:                 422   AIC:                                      1714.
Df Residuals:                     407   BIC:                                      1774.
Df Model:                          15                                                  
Covariance Type:            nonrobust                                                  
=======================================================================================
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
Explanatory Var #1      1.3078      0.013    102.159      0.000       1.283       1.333
Explanatory Var #2      1.7667      0.009    203.101      0.000       1.750       1.784
Explanatory Var #3      6.6083      0.177     37.232      0.000       6.259       6.957
Explanatory Var #4      2.0725      0.011    189.896      0.000       2.051       2.094
Explanatory Var #5     -0.7776      0.011    -68.372      0.000      -0.800      -0.755
Explanatory Var #6      0.0124      0.008      1.595      0.111      -0.003       0.028
Explanatory Var #7      0.0286      0.006      4.890      0.000       0.017       0.040
Explanatory Var #8      0.3510      0.080      4.398      0.000       0.194       0.508
Explanatory Var #9     -0.0293      0.010     -2.838      0.005      -0.050      -0.009
Explanatory Var #10     0.1292      0.011     11.409      0.000       0.107       0.152
Explanatory Var #11     0.0145      0.013      1.152      0.250      -0.010       0.039
Explanatory Var #12     0.0752      0.009      8.698      0.000       0.058       0.092
Explanatory Var #13     0.7414      0.177      4.198      0.000       0.394       1.089
Explanatory Var #14    -0.0218      0.011     -1.974      0.049      -0.043   -8.61e-05
Explanatory Var #15     0.1212      0.012     10.324      0.000       0.098       0.144
==============================================================================
Omnibus:                        0.087   Durbin-Watson:                   1.875
Prob(Omnibus):                  0.957   Jarque-Bera (JB):                0.188
Skew:                          -0.004   Prob(JB):                        0.910
Kurtosis:                       2.897   Cond. No.                         287.
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Dependent Variable: The dependent variable is referred to as "Dependent Var," and the analysis aims to explain its variation based on the independent variables.

R-squared: The R-squared value is a measure of how well the independent variables explain the variation in the dependent variable. In this case, the uncentered R-squared is very high at 0.999, indicating that the independent variables together explain 99.9% of the variation in the dependent variable.

Adj. R-squared: The adjusted R-squared takes into account the number of independent variables in the model. In this case, the uncentered adjusted R-squared is also very high at 0.999.

F-statistic: The F-statistic is used to test the overall significance of the regression model. A high F-statistic, in this case, 4.509e+04, suggests that the model as a whole is statistically significant.

Prob (F-statistic): The probability associated with the F-statistic is very low (0.00), indicating that the overall model is highly statistically significant.

No. Observations: There are 422 observations in the dataset.

Degrees of Freedom (Df Residuals, Df Model):

Df Residuals: 407 degrees of freedom for the residuals.

Df Model: 15 degrees of freedom for the model.

Covariance Type: The analysis was performed using a non-robust covariance type.

Coefficients (coef): The coefficients for each of the 15 explanatory variables (independent variables) are provided, along with their standard errors.

t-statistic (t): The t-statistic measures the significance of each coefficient. A higher absolute t-value indicates greater significance.

P>|t| (p-value): This represents the probability that the coefficient is not significantly different from zero. Low p-values (typically < 0.05) suggest that the corresponding variable is statistically significant in explaining the dependent variable.
[0.025 0.975]: These values provide a 95% confidence interval for each coefficient.

Omnibus: The Omnibus test is a test of the normality of the residuals. A higher value is generally better, and a p-value close to 1 indicates that the residuals are normally distributed.

Durbin-Watson: This statistic tests for the presence of autocorrelation (correlation of residuals at different points in time). A value close to 2 (in this case, 1.875) is often considered desirable as it suggests little autocorrelation.

Jarque-Bera (JB): The Jarque-Bera test also assesses the normality of residuals.

Skew: Measures the skewness of the residuals. A value close to 0 suggests a symmetric distribution.

Kurtosis: Measures the kurtosis of the residuals. A value of 2.897 indicates the shape of the distribution of residuals.

Cond. No.: The condition number (287) is a measure of multicollinearity, with higher values indicating a potential issue with multicollinearity among the independent variables.

In summary, this regression analysis appears to have a very high R-squared value, indicating that the independent variables explain a significant portion of the variation in the dependent variable. The F-statistic and associated p-value suggest that the model is highly significant. The individual coefficients and their p-values allow you to assess the significance of each independent variable in explaining the dependent variable. Additionally, the diagnostic tests (Omnibus, Durbin-Watson, Jarque-Bera, Skew, and Kurtosis) provide insights into the quality of the model and the distribution of residuals.

## Part B

1.	Use error values from the OLS model to calculate their standard deviation and autocorrelation values for the first three lags. 

In [144]:
pred_vals = model_reg.predict(X)
errors = y - pred_vals
errors

0      4.825713
1     -0.197803
2      0.975013
3      1.853953
4      1.110946
         ...   
417   -3.236116
418   -0.454778
419    2.851672
420   -2.647010
421    1.035218
Length: 422, dtype: float64

In [167]:
st_dev = np.std(errors)
st_dev

1.7757282184664

In [146]:
lags = 3
autocorrelation = [np.corrcoef(df[i:], df[:-i])[0, 1] for i in range(1, lags + 1)]
autocorrelation

[0.9338889427200325, 0.8568714054990086, 0.8841512562268686]

2.	Then, run the GLS model accordingly. 

get from lesson 3

use cov_matrix function (1, cov1, cov2, cov3, np.zeros(418))

In [147]:
gls_model = sm.GLS(y, X)

In [148]:
order = 1
cov_structure = sm.cov_struct.Autoregressive(order)
gls_model.cov_struct = cov_structure

c:\Users\johns\AppData\Local\Programs\Python\Python312\Lib\site-packages\statsmodels\genmod\cov_struct.py:796: FutureWarning: grid=True will become default in a future version
  warnings.warn(


In [149]:
gls_results = gls_model.fit()

3.	Show your summary table in Python and interpret your results in the summary report.

In [150]:
print(gls_results.summary())

                                 GLS Regression Results                                
Dep. Variable:          Dependent Var   R-squared (uncentered):                   0.999
Model:                            GLS   Adj. R-squared (uncentered):              0.999
Method:                 Least Squares   F-statistic:                          4.509e+04
Date:                Mon, 23 Oct 2023   Prob (F-statistic):                        0.00
Time:                        21:20:27   Log-Likelihood:                         -841.76
No. Observations:                 422   AIC:                                      1714.
Df Residuals:                     407   BIC:                                      1774.
Df Model:                          15                                                  
Covariance Type:            nonrobust                                                  
                          coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------

Dependent Variable: The dependent variable is labeled as "Dependent Var," and the analysis aims to explain its variation based on the independent variables.

R-squared (uncentered): The uncentered R-squared is very high at 0.999, indicating that the independent variables, as a whole, explain 99.9% of the variation in the dependent variable. However, note that this R-squared is computed without centering because the model does not contain a constant.

Adj. R-squared (uncentered): The uncentered adjusted R-squared is also very high at 0.999. This value takes into account the number of independent variables in the model while adjusting the R-squared.

F-statistic: The F-statistic is used to test the overall significance of the GLS regression model. A high F-statistic, in this case, 4.509e+04, suggests that the model as a whole is statistically significant.

Prob (F-statistic): The probability associated with the F-statistic is very low (0.00), indicating that the overall GLS model is highly statistically significant.

No. Observations: There are 422 observations in the dataset.

Degrees of Freedom (Df Residuals, Df Model):

Df Residuals: 407 degrees of freedom for the residuals.

Df Model: 15 degrees of freedom for the model.

Covariance Type: The analysis was performed using a nonrobust covariance type.

Coefficients (coef): The coefficients for each of the 15 explanatory variables (independent variables) are provided, along with their standard errors.

t-statistic (t): The t-statistic measures the significance of each coefficient. A higher absolute t-value indicates greater significance.

P>|t| (p-value): This represents the probability that the coefficient is not significantly different from zero. Low p-values (typically < 0.05) suggest that the corresponding variable is statistically significant in explaining the dependent variable.


In summary, this GLS regression analysis indicates a highly significant model with a very high R-squared value. The coefficients and their associated statistics provide information about the individual explanatory variables' contributions to explaining the dependent variable. The provided information is quite similar to a standard OLS regression output, with the key difference being the use of GLS regression instead of OLS.

## Part C

1.	Split the dataset into two as the training and test sets (test size = 0.5). 

In [151]:
from sklearn.model_selection import train_test_split

In [152]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=42)

2.	Run the Lasso model with alpha=1 and estimate the coefficients using the training set. 

In [153]:
model_lasso = lm.Lasso(alpha=1).fit(X_train,y_train)
model_lasso.coef_

array([ 1.26972628,  1.68394638,  2.02626245,  2.08756512, -0.91746375,
       -0.        ,  0.        , -0.        ,  0.        , -0.        ,
        0.01314162,  0.        , -0.        ,  0.        , -0.03617731])

3.	Then, calculate the mean absolute percentage error using the test set. 

Method 1

In [154]:
predictions = model_lasso.predict(X_test)

In [155]:
absolute_percentage_errors = abs((y_test - predictions) / y_test) * 100

In [156]:
mape = absolute_percentage_errors.mean()    

In [157]:
print("Mean Absolute Percentage Error (MAPE):", mape)

Mean Absolute Percentage Error (MAPE): 4.432190198291583


Method 2

In [158]:
mape = mean_absolute_percentage_error(y_test, predictions)
mape = mape*100
print("Mean Absolute Percentage Error (MAPE):", mape)

Mean Absolute Percentage Error (MAPE): 4.432190198291583


4.	Find an approximate value for alpha that minimizes the mean absolute percentage error.

In [159]:
model_lasso = lm.Lasso(alpha=0.01).fit(X_train,y_train)
model_lasso.coef_

predictions = model_lasso.predict(X_test)
absolute_percentage_errors = abs((y_test - predictions) / y_test) * 100
mape = absolute_percentage_errors.mean()    
print("Mean Absolute Percentage Error (MAPE) for Alpha 0.01:", mape)

Mean Absolute Percentage Error (MAPE) for Alpha 0.01: 0.04429881163345532


In [160]:
model_lasso = lm.Lasso(alpha=0.1).fit(X_train,y_train)
model_lasso.coef_

predictions = model_lasso.predict(X_test)
absolute_percentage_errors = abs((y_test - predictions) / y_test) * 100
mape = absolute_percentage_errors.mean()    
print("Mean Absolute Percentage Error (MAPE) for Alpha 0.1:", mape)

Mean Absolute Percentage Error (MAPE) for Alpha 0.1: 0.44323186549579874


In [161]:
model_lasso = lm.Lasso(alpha=10).fit(X_train,y_train)
model_lasso.coef_

predictions = model_lasso.predict(X_test)
absolute_percentage_errors = abs((y_test - predictions) / y_test) * 100
mape = absolute_percentage_errors.mean()    
print("Mean Absolute Percentage Error (MAPE) for Alpha 10:", mape)

Mean Absolute Percentage Error (MAPE) for Alpha 10: 8.587094351590437


In [162]:
model_lasso = lm.Lasso(alpha=100).fit(X_train,y_train)
model_lasso.coef_

predictions = model_lasso.predict(X_test)
absolute_percentage_errors = abs((y_test - predictions) / y_test) * 100
mape = absolute_percentage_errors.mean()    
print("Mean Absolute Percentage Error (MAPE) for Alpha 100:", mape)

Mean Absolute Percentage Error (MAPE) for Alpha 100: 44.287702093738424


Mean Absolute Percentage Error (MAPE) for Alpha 0.01: 0.04429881163345532

Mean Absolute Percentage Error (MAPE) for Alpha 0.1: 0.44323186549579874

Mean Absolute Percentage Error (MAPE) for Alpha 1: 4.432190198291583

Mean Absolute Percentage Error (MAPE) for Alpha 10: 8.587094351590437

Mean Absolute Percentage Error (MAPE) for Alpha 100: 44.287702093738424

## Part D

1.	Use the demand data given in the table and develop an appropriate forecasting model (i.e., the tailored regularization discussed in the class—see your slides for more info) that exploits the available information given in the table as much as possible.

In [163]:
month = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25]
Demand = [100, 112, 107, 103, 91, 85, 84, 85, 79, 81, 134, 86, 99, 89, 111, 114, 118, 163, 193, 143, 144, 202, 158, 160, 144]
Adv_Demand = [71, 30, 75, 64, 41, 51, 42, 51, 57, 49, 134, 52, 99, 56, 81, 79, 73, 163, 193, 99, 91, 202, 105, 101, 96]

In [164]:
forecast = pd.DataFrame({'month': month, 'Demand': Demand, 'Adv_Demand': Adv_Demand})
forecast

month  Demand  Adv_Demand
0       1     100          71
1       2     112          30
2       3     107          75
3       4     103          64
4       5      91          41
5       6      85          51
6       7      84          42
7       8      85          51
8       9      79          57
9      10      81          49
10     11     134         134
11     12      86          52
12     13      99          99
13     14      89          56
14     15     111          81
15     16     114          79
16     17     118          73
17     18     163         163
18     19     193         193
19     20     143          99
20     21     144          91
21     22     202         202
22     23     158         105
23     24     160         101
24     25     144          96

In [165]:
# Sample data for your DataFrame
month = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25]
Demand = [100, 112, 107, 103, 91, 85, 84, 85, 79, 81, 134, 86, 99, 89, 111, 114, 118, 163, 193, 143, 144, 202, 158, 160, 144]
Adv_Demand = [71, 30, 75, 64, 41, 51, 42, 51, 57, 49, 134, 52, 99, 56, 81, 79, 73, 163, 193, 99, 91, 202, 105, 101, 96]

# Create a DataFrame with the given data
data = {
    'month': month,
    'Demand': Demand,
    'Adv_Demand': Adv_Demand,
}
forecast = pd.DataFrame(data)

# Prepare the data
X = forecast[['month', 'Adv_Demand']]  # Adjust the column names accordingly
Y = forecast['Demand']

# Initialize and fit a Ridge regression model
alpha = 1.0  # You can adjust the regularization strength as needed
ridge_model = Ridge(alpha=alpha)
ridge_model.fit(X, Y)

# Predict future demand
future_X = np.array([[6, 115]])  # Adjust these values for your specific forecast
forecasted_demand = ridge_model.predict(future_X)

print("Forecasted Demand:", forecasted_demand)

Forecasted Demand: [126.09920919]


c:\Users\johns\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\base.py:465: UserWarning: X does not have valid feature names, but Ridge was fitted with feature names
  warnings.warn(


The result "Forecasted Demand: [149.88505747]" is the output of your Ridge regression model, representing the forecasted or predicted demand for a specific set of input variables. 

Forecasted Demand: This is the predicted value for the demand, which is what we are trying to forecast. In this case, the model estimates that the demand will be approximately 149.89 units.

Interpretation: The forecasted demand of approximately 149.89 units is based on the input variables used to make the prediction, which are 'month' and 'Adv_Demand' in this case. This means that, given a 'month' value of 6 and an 'Adv_Demand' value of 115 (as specified in your 'future_X' array), the model predicts that the demand will be around 149.89 units.

Confidence: The forecasted value is not an absolute truth but a statistical estimate. The actual demand may vary, and the accuracy of the prediction depends on the quality of the data, the model's assumptions, and other factors. To assess the reliability of the forecast, we will use metrics like mean squared error (MSE) or root mean squared error (RMSE) to measure the model's prediction accuracy when compared to actual historical data.

In summary, the result "Forecasted Demand: [149.88505747]" is the model's best estimate of the demand.


2.	Interpret your results.

In [166]:
# Create a DataFrame for the data
data = pd.DataFrame({'Month':     [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25],
                    'Demand':     [100, 112, 107, 103, 91, 85, 84, 85, 79, 81, 134, 86, 99, 89, 111, 114, 118, 163, 193, 143, 144, 202, 158, 160, 144],
                    'Adv_Demand': [71, 30, 75, 64, 41, 51, 42, 51, 57, 49, 134, 52, 99, 56, 81, 79, 73, 163, 193, 99, 91, 202, 105, 101, 96]})

# Machine Learning Model
X = data[['Month', 'Adv_Demand']]
y = data['Demand']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Lasso Regularization
lasso_model = Lasso(alpha=0.01)
lasso_model.fit(X_train, y_train)

# Ridge Regularization
ridge_model = Ridge(alpha=0.01)
ridge_model.fit(X_train, y_train)

# Evaluate the models
lasso_predictions = lasso_model.predict(X_test)
ridge_predictions = ridge_model.predict(X_test)

# Calculate evaluation metrics
lasso_rmse = sqrt(mean_squared_error(y_test, lasso_predictions))
ridge_rmse = sqrt(mean_squared_error(y_test, ridge_predictions))

# Print RMSE for each model
print(f'Lasso RMSE: {lasso_rmse}')
print(f'Ridge RMSE: {ridge_rmse}')


Lasso RMSE: 13.674863351909064
Ridge RMSE: 13.674350201193812
